### ISSUE TITLE
SkypeAuthException ("Couldn't retrieve t field from login response", <Response [200]>) when trying o login.

### ISSUE THREAD

**Issue by pavlovmarch** [OllieTerrance/SkPy#65]
I get this error when I am trying to login:
SkypeAuthException at /index/
("Couldn't retrieve t field from login response", <Response [200]>)

Also, when I run the code on my local machine it works OK. 
But when I try to run it on server it fails with error above.

It works OK with few Skype accounts. 
But I get such error on Live accounts and another Skype accounts.

On Skype for Web I can login fine, no captcha or other problems.

**Comment by pavlovmarch** [comment]
I am using 0.8 version of SkPy

**Comment by OllieTerrance** [comment]
Presumably you are using Skype for Web from your local machine?  You may be being blocked from your server's IP if you logged in elsewhere recently.  If you can proxy Skype for Web via your server, you may be able to resolve it with a captcha.

**Comment by pavlovmarch** [comment]
Thanks for your reply!
But what if I get this error: SkypeApiException: ('404 response from PUT https://client-s.gateway.me
ssenger.live.com/v1/users/ME/endpoints/bee77efb-7a06-692a-aafd-97a003fd126c/pres
enceDocs/messagingService', <Response [404]>)

From the same server for some users.

**Comment by pavlovmarch** [comment]
I receive this error using live accounts.

**Comment by pavlovmarch** [comment]
Is it possible to change host adress ?

**Comment by OllieTerrance** [comment]
> But what if I get this error: SkypeApiException: ('404 response from PUT https://client-s.gateway.me
ssenger.live.com/v1/users/ME/endpoints/bee77efb-7a06-692a-aafd-97a003fd126c/pres
enceDocs/messagingService', <Response [404]>)

That one's been discussed at length in #47 and #54 -- a few people have reported it, but it tends to start working after a while.  If you can provide any more info than what's in the previous issues, we can try and debug it further.

> Is it possible to change host adress ?

You can't change your own IP besides running on a different server, nor the API hostname as Skype tells you which host to use (and bounces you away from the wrong ones).

**Comment by pavlovmarch** [comment]
Looks like it works on account when I run it on local machine, but it falls when I run it on server (I get 404 error). So I think the ploblem is in that I run it from different regions.

**Comment by rooterkyberian** [comment]
It seems there is additional 2 factor verification prompt which is causing this error.

**Comment by yf33** [comment]
after login web skype and  try again

**Comment by olegspk** [comment]
skpy module does not work will fix?

**Comment by Ruined1** [comment]
> after login web skype and try again

This WORKS, and I think I understand why, but is there a way to circumvent this issue in the code? Or it is just a "deal with it" situation?

**Comment by Terrance** [comment]
If it's a captcha, unlikely.  If it's an interstitial page asking about your account (e.g. confirming recovery information, terms of service changes), then it's not this library's business to read or accept such prompts.

The solution is to sign in in a browser and accept such prompts first, there's not much else that can be done here.  Attempting to handle them feels out of scope here given they apply to Microsoft accounts rather than Skype specifically.

**Comment by Ruined1** [comment]
> If it's a captcha, unlikely. If it's an interstitial page asking about your account (e.g. confirming recovery information, terms of service changes), then it's not this library's business to read or accept such prompts.
> 
> The solution is to sign in in a browser and accept such prompts first, there's not much else that can be done here. Attempting to handle them feels out of scope here given they apply to Microsoft accounts rather than Skype specifically.

It literally asked for no additional information. I provided my username and password to skype for web, and it signed me in. I then closed the page, tried it again, and no longer received the "Couldn't retrieve t field from login response" error, if that helps any.

**Comment by Terrance** [comment]
It's possibly some form of bot detection (think e.g. NoCaptcha ReCaptcha) that passes a challenge in a browser with JavaScript, in which case we may be out of luck.

Past sending the username and password, the responses all tend to be a spartan HTML page with a number of hidden form fields and some JavaScript.  If there's a `t` field present then we have the token we need to complete the auth; if not then it's pending some other action, but it's unclear what any given page wants unless someone can reproduce it in a browser.

**Comment by namnguyendsn** [comment]
> It's possibly some form of bot detection (think e.g. NoCaptcha ReCaptcha) that passes a challenge in a browser with JavaScript, in which case we may be out of luck.
> 
> Past sending the username and password, the responses all tend to be a spartan HTML page with a number of hidden form fields and some JavaScript. If there's a `t` field present then we have the token we need to complete the auth; if not then it's pending some other action, but it's unclear what any given page wants unless someone can reproduce it in a browser.

hi Terrance.
thank for your lib.
after execute line 211 in conn.py file
resp = self.sess.request(method, url, headers=headers, **kwargs)
i printed resp.text and see below value:
`<html><head><noscript>JavaScript required to sign in<meta http-equiv="Refresh" content="0; URL=https://login.live.com/jsDisabled.srf?mkt=EN-US&lc=1033&uaid=d4219c9d226c4301aa3ad1e95265fa47&client_flight=ReservedFlight33,ReservedFligh"/></noscript><title>Continue</title><script type="text/javascript">function OnBack(){}function OnNext(){}function DoSubmit(){var subt=false;if(!subt){subt=true;document.fmHF.submit();}}</script></head><body onload="javascript:DoSubmit();"><form name="fmHF" id="fmHF" action="https://account.live.com/identity/confirm?mkt=EN-US&uiflavor=web&id=293290&ru=https://login.live.com/login.srf%3fid%3d293290%26opid%3d24F7983BC963BB6D%26opidt%3d1564413537" method="post" target="_top"><input type="hidden" name="client_flight" id="client_flight" value="ReservedFlight33,ReservedFligh"><input type="hidden" name="ipt" id="ipt" value="DQG*1ke18uL69MMGPmA3zfuwtSmWz6bYozwzvFlrbkIus3bnLpR8LTGPfFqdGPbahu2sExRFXj8ZaioRXpwqV2TflabhkgIX1gZXsf6RXhiLUOJke226HzDxwrk4hPDkBxRNSQRAEqaHoTRgOBQy*x7TIOLqWmrbAwrqusujxm1xv16Ydnua4p62zLdZ5iu9PWrlbxfWlYMBwA39TG77qquBx!MQpAORVIv7YrUwROEFhNj1ygtBRaxveQfecZHe0KCZNd9wq15AJwd64rE5Ba*1nYWfN6GrDhJrFJXXsbF7tSEAuxVpy455H*omgIsgYQ41wFDZvtE3iuMlOJ3ihOV7P5rCdSFJPPD5hfBqjGTlniR6T7yUgOQ8cUyfZxQdri7cUk!*k944gpKpqP0FeIUQzk4uCIA2XsJnKn7OGKS6n7vg0KdYjd1KaUpPyuE4ivXAmC2Y8svAdPpOOcTnGvxgV8x7LhNl4mvUE8FbuRMR"><input type="hidden" name="pprid" id="pprid" value="29b03d9f0f0ab51e"><input type="hidden" name="uaid" id="uaid" value="d4219c9d226c4301aa3ad1e95265fa47"></form></body></html>`

it require Javascript for login process.  how to pass over?
sorry for my bad english

thank you

**Comment by piropiro2018** [comment]
My environment is CentOS7 and I got into a similar situation.
Added proxies param,It seems that the proxy specification,and User-Agent is valid.
I　sent　message,from server.

`
headers["User-Agent"] = 'User agent of the browser that logged in skype for web of Windows or MAC'
proxyDict = { 
              "http"  : http_proxy, 
              "https" : https_proxy
 }`

`resp = self.sess.request(method, url, headers=headers, proxies=proxyDict, **kwargs)`


I specified a proxy that is in the same country.（logged in skype for web of Windows or MAC）

Along the way, there was a warning like this, but the message was sent

{"errorCode":752,"message":"User is in a different cloud. See 'Location' header for users current cloud."}

**Comment by Terrance** [comment]
> Added proxies param

You don't need to edit the code for that, see #120.

> there was a warning like this

That's expected and should be handled for you -- SkPy connects to a default messaging host, but will follow Skype's redirects to a local one.

**Comment by flannan-h** [comment]
Hey everyone,

Recognize that this 2FA issue is unavoidable and usually solved by logging in & out of web skype, as @yf33 pointed out.

Still, I'm curious to see if anyone else has experienced an increase in failed logins recently. Usually, my skpy scripts will run hourly for 2+ months before requiring a manual login. Recently, the failure frequency has severely increased and feels more random.

For example, in a script that attempts three consecutive logins, the first and third log-ins may succeed while the second one fails with the 't field' error. No AuthLimit reached. So it now seems independent of any 2FA/manual logins. Fearful of a more stringent underlying bot detection as @Terrance suggested. FWIW, this new performance issue has been replicated several times within our org

Looking forward to hearing any other similar experiences #LongLiveSkpy

**Comment by starraw** [comment]
hi is there a new update in skype or skpy lib? suddenly having issue with login in to skype.

was using it without any error all along until today. try different account also facing the same error. I have no issue login into skype via web online as well.

File "C:\Program Files\Python\Python38-32\lib\site-packages\skpy\main.py", lin
e 69, in __init__
    self.conn.getSkypeToken()
  File "C:\Program Files\Python\Python38-32\lib\site-packages\skpy\conn.py", lin
e 266, in getSkypeToken
    self.liveLogin(user, pwd)
  File "C:\Program Files\Python\Python38-32\lib\site-packages\skpy\conn.py", lin
e 368, in liveLogin
    self.tokens["skype"], self.tokenExpiry["skype"] = SkypeLiveAuthProvider(self
).auth(user, pwd)
  File "C:\Program Files\Python\Python38-32\lib\site-packages\skpy\conn.py", lin
e 542, in auth
    t = self.sendCreds(user, pwd, params)
  File "C:\Program Files\Python\Python38-32\lib\site-packages\skpy\conn.py", lin
e 575, in sendCreds
    raise SkypeAuthException(errMsg, loginResp)
skpy.core.SkypeAuthException: ("Couldn't retrieve t field from login response",
<Response [200]>)

anyone facing the same issue as well?

**Comment by kozlou** [comment]
> hi is there a new update in skype or skpy lib? suddenly having issue with login in to skype.
> anyone facing the same issue as well?

I am facing this too. Started from today. When you login in a skype web, there is a message about changed terms of service or something. I've clicked accept and logged in successfuly, but my bot doesn't :(

**Comment by hmcvieira** [comment]
> hi is there a new update in skype or skpy lib? suddenly having issue with login in to skype.
> 
> was using it without any error all along until today. try different account also facing the same error. I have no issue login into skype via web online as well.
> 
> File "C:\Program Files\Python\Python38-32\lib\site-packages\skpy\main.py", lin
> e 69, in **init**
> self.conn.getSkypeToken()
> File "C:\Program Files\Python\Python38-32\lib\site-packages\skpy\conn.py", lin
> e 266, in getSkypeToken
> self.liveLogin(user, pwd)
> File "C:\Program Files\Python\Python38-32\lib\site-packages\skpy\conn.py", lin
> e 368, in liveLogin
> self.tokens["skype"], self.tokenExpiry["skype"] = SkypeLiveAuthProvider(self
> ).auth(user, pwd)
> File "C:\Program Files\Python\Python38-32\lib\site-packages\skpy\conn.py", lin
> e 542, in auth
> t = self.sendCreds(user, pwd, params)
> File "C:\Program Files\Python\Python38-32\lib\site-packages\skpy\conn.py", lin
> e 575, in sendCreds
> raise SkypeAuthException(errMsg, loginResp)
> skpy.core.SkypeAuthException: ("Couldn't retrieve t field from login response",
> <Response [200]>)
> 
> anyone facing the same issue as well?

I am facing this too. Recently there have been an unusual increasy in failed login attempts. But todays it is impossible to login, tried already with multiples accounts in multiple computers. Always goot the same error message 
SkypeAuthException: ("Couldn't retrieve t field from login response", <Response [200]>)

**Comment by Terrance** [comment]
Looks like Microsoft have changed something with the login page, as similar issues are being reported with [skype4pidgin](https://github.com/EionRobb/skype4pidgin/issues/656) and Miranda-NG which also use Skype for Web's APIs.

**Comment by starraw** [comment]
> Looks like Microsoft have changed something with the login page, as similar issues are being reported with [skype4pidgin](https://github.com/EionRobb/skype4pidgin/issues/656) and Miranda-NG which also use Skype for Web's APIs.

Hi Terrance, thanks for notifying. 
Does anyone has any way to resolve it? really will appreciate any kind help.

**Comment by verybasicneeds** [comment]
Hello,

Same problem here, it seems to work randomly after xx attemps.

**Comment by starraw** [comment]
Anyone else also facing this issue and able to resolve it? In need of help here.. =(

**Comment by kozlou** [comment]
> Anyone else also facing this issue and able to resolve it? In need of help here.. =(

It seems like the problem is in auth mechanishm that was recently changed by microsoft. So we need some superhero like @Terrance to solve this

**Comment by dukv2001** [comment]
Come on, super hero @Terrance we need you

**Comment by Terrance** [comment]
The [`soap-login` branch](https://github.com/Terrance/SkPy/tree/soap-login) adds support for SOAP-based authentication.  This is exposed internally as `SkypeSOAPAuthProvider`, though the higher-level authentication methods will already call this instead of the previous flow, so no code changes should be needed.

Your Skype account must be linked to a Microsoft account in order for SOAP logins to succeed.  It looks like this also supports two-factor-enabled accounts if you provide an application-specific password.

Please comment in here if this branch works for you or not.

**Comment by flannan-h** [comment]
@Terrance new branch just worked for me! Have tried a few separate login attempts, all successful. Will keep posted with any updates - thanks a million!

**Comment by kozlou** [comment]
> Please comment in here if this branch works for you or not.
Yeah! It works!

**Comment by starraw** [comment]
hi @Terrance 
My alternate works! but another account is having the profile is accrual error. any idea what is causing that. and really appreciate your help in solving the problem! at least i can use the alternate for the time being!

**Comment by dukv2001** [comment]
![image](https://user-images.githubusercontent.com/61406754/90806497-ae097b80-e34f-11ea-9794-f4f943bcfd17.png)

**Comment by dukv2001** [comment]
[f.txt](https://github.com/Terrance/SkPy/files/5104613/f.txt)

**Comment by dukv2001** [comment]
[kc.v.txt](https://github.com/Terrance/SkPy/files/5104614/kc.v.txt)

**Comment by dukv2001** [comment]
![image](https://user-images.githubusercontent.com/61406754/90806719-ff196f80-e34f-11ea-8833-2a2d61a130da.png)

**Comment by dukv2001** [comment]
He still doesn't support

**Comment by dukv2001** [comment]
He is working normally, thank you

**Comment by hmcvieira** [comment]
It is working normally, even without the new branch.

**Comment by Terrance** [comment]
The old authentication flow continues not to work for me, and I think in any case it's worth switching to gain support for 2FA-enabled accounts.  I'm interested if anyone can authenticate using the existing flow on `master` but not the new implementation in `soap-login`.  If it works consistently then I'll get an updated release out.

**Comment by flannan-h** [comment]
> The old authentication flow continues not to work for me, and I think in any case it's worth switching to gain support for 2FA-enabled accounts.

Same and agreed!

**Comment by pbrito55** [comment]
Some users for me work others do not with SOAP login.

The strange thing is that  if I use CURL with the SOAP XML template I get the token always  for all users if I use requests lib of skpy soap it fails. Only for one is ok.

curl -v -X POST -H "Content-Type: text/xml" --data "@temp.xml"  https://login.live.com/RST.srf

with temp.xml 
<Envelope xmlns='http://schemas.xmlsoap.org/soap/envelope/'
       xmlns:wsse='http://schemas.xmlsoap.org/ws/2003/06/secext'
       xmlns:wsp='http://schemas.xmlsoap.org/ws/2002/12/policy'
       xmlns:wsa='http://schemas.xmlsoap.org/ws/2004/03/addressing'
       xmlns:wst='http://schemas.xmlsoap.org/ws/2004/04/trust'
       xmlns:ps='http://schemas.microsoft.com/Passport/SoapServices/PPCRL'>
       <Header>
           <wsse:Security>
               <wsse:UsernameToken Id='user'>
                   <wsse:Username>xxxxxxxx@outlook.com</wsse:Username>
                   <wsse:Password>xxxxxxxxxxx</wsse:Password>
               </wsse:UsernameToken>
           </wsse:Security>
       </Header>
       <Body>
           <ps:RequestMultipleSecurityTokens Id='RSTS'>
               <wst:RequestSecurityToken Id='RST0'>
                   <wst:RequestType>http://schemas.xmlsoap.org/ws/2004/04/security/trust/Issue</wst:RequestType>
                   <wsp:AppliesTo>
                       <wsa:EndpointReference>
                           <wsa:Address>wl.skype.com</wsa:Address>
                       </wsa:EndpointReference>
                   </wsp:AppliesTo>
                   <wsse:PolicyReference URI='MBI_SSL'></wsse:PolicyReference>
               </wst:RequestSecurityToken>
           </ps:RequestMultipleSecurityTokens>
       </Body>
    </Envelope>

always returns the token

**Comment by Terrance** [comment]
> if I use requests lib of skpy soap it fails

@pbrito55 In what way?

**Comment by pbrito55** [comment]
`<?xml version="1.0" encoding="utf-8" ?><S:Envelope xmlns:S="http://schemas.xmlsoap.org/soap/envelope/"><S:Header></S:Header><S:Body><wst:RequestSecurityTokenResponseCollection xmlns:S="http://schemas.xmlsoap.org/soap/envelope/" xmlns:wst="http://schemas.xmlsoap.org/ws/2004/04/trust" xmlns:wsse="http://schemas.xmlsoap.org/ws/2003/06/secext" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" xmlns:saml="urn:oasis:names:tc:SAML:1.0:assertion" xmlns:wsp="http://schemas.xmlsoap.org/ws/2002/12/policy" xmlns:psf="http://schemas.microsoft.com/Passport/SoapServices/SOAPFault"><wst:RequestSecurityTokenResponse><wsp:AppliesTo xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/03/addressing"><wsa:EndpointReference><wsa:Address>wl.skype.com</wsa:Address></wsa:EndpointReference></wsp:AppliesTo><S:Fault><faultcode>wsse:FailedAuthentication</faultcode><faultstring>Authentication Failure</faultstring><psf:pp><psf:reqstatus>0x800488ff</psf:reqstatus><psf:errorstatus>0x80049c2a</psf:errorstatus><psf:flowurl>https://login.live.com/login.srf?lc=1033&amp;st=-DbelVS3IredQ1!RG6astrqxfm8UT0NwDep62qrTN2PHtoMDRHt*ecwBJniclDylAcLLE46rkFczXZ*hf*MpPXSqmlBkWvA8JGQDHdtZHcFM8vFQ8FCgB85*DIaf*Ad3o5UCwtJZhRt0YLKjuY6KReZdFt1hZDMvLUo4*j6rm2sA9sfJ!h8Kdnr0lLT13zzFcHaODVGAGeDjIKJB8Nfij2uTmVUSwJ2v23ZRB1o*5uD0U&amp;seclog=10&amp;ppsit=2&amp;aplcfi=1</psf:flowurl></psf:pp></S:Fault></wst:RequestSecurityTokenResponse></wst:RequestSecurityTokenResponseCollection></S:Body></S:Envelope>`

**Comment by pbrito55** [comment]
this is the same answer I get back with request or from here
`loginResp = self.conn("POST", "{0}/RST.srf".format(SkypeConnection.API_MSACC),
                              data=self.template.format(self.encode(user), self.encode(pwd)))
       loginResp.text`

**Comment by verybasicneeds** [comment]
it works for me with the soap-login branch, many thanks

**Comment by Terrance** [comment]
@pbrito55 Does it start working if you add `headers={"Content-Type": "text/xml"}` to the `self.conn()` call?  It doesn't seem to be needed for others, but I'm not sure what else differs here.

**Comment by pbrito55** [comment]
I still get the loginResp.text
'<?xml version="1.0" encoding="utf-8" ?><S:Envelope xmlns:S="
http://schemas.xmlsoap.org/soap/envelope/"><S:Header></S:Header><S:Body><wst:RequestSecurityTokenResponseCollection
xmlns:S="http://schemas.xmlsoap.org/soap/envelope/" xmlns:wst="
http://schemas.xmlsoap.org/ws/2004/04/trust" xmlns:wsse="
http://schemas.xmlsoap.org/ws/2003/06/secext" xmlns:wsu="
http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd"
xmlns:saml="urn:oasis:names:tc:SAML:1.0:assertion" xmlns:wsp="
http://schemas.xmlsoap.org/ws/2002/12/policy" xmlns:psf="
http://schemas.microsoft.com/Passport/SoapServices/SOAPFault"><wst:RequestSecurityTokenResponse><wsp:AppliesTo
xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/03/addressing
"><wsa:EndpointReference><wsa:Address>wl.skype.com</wsa:Address></wsa:EndpointReference></wsp:AppliesTo><S:Fault><faultcode>wsse:FailedAuthentication</faultcode><faultstring>Authentication
Failure</faultstring><psf:pp><psf:reqstatus>0x800488ff</psf:reqstatus><psf:errorstatus>0x80049c2a</psf:errorstatus><psf:flowurl>
https://login.live.com/login.srf?lc=1033&amp;st=-Dccd6T!1UAvu5reVJ0bBevR4CQgKyMjf!lzsBUTHCBTzm7ABhR7QbSFWvDGjogj!gJPzxeFsPV4l78WGhQIOulYaamEnBbqIlZU!IW1yqOf69UAfaM*LdQrAYQaTOryvb8!4ODCkcMFcGu75OjHfbwU48bVymVM6CdFg652j3auU!!dVRUcvFm*ZM0g2NEusYa49xGs5Z!8EQIgI8a0MvyASWMdmRV!EYmFZYOj3I9c!&amp;seclog=10&amp;ppsit=2&amp;aplcfi=1</psf:flowurl></psf:pp></S:Fault></wst:RequestSecurityTokenResponse></wst:RequestSecurityTokenResponseCollection></S:Body></S:Envelope>
'

**Comment by pbrito55** [comment]
all working it had to something stupid... proxys ..... thanks //PB

**Comment by TemurKhan** [comment]
By using new branch as mentioned by @Terrance i were unable to login as I were using just Phone number for skype account. 
I attached a microsoft  email with phone number, now its working fine. Thanks

**Comment by jacobdenis** [comment]
Hi, I already downloaded the new update. but still, experience the same issue.  

here is my connection maybe there is something wrong with what I am doing thank you.
![image](https://user-images.githubusercontent.com/28717802/90996996-d7960180-e5f2-11ea-9d11-3fd82aed7ada.png)

**Comment by Terrance** [comment]
If you're getting the same error message from the title of this issue then you're not on the new update -- the authentication was rewritten to use different APIs producing different error messages.

You're also reimplementing `Skype('username', 'password', tokenFile='.tokens-app')`, which does all of what you just quoted.

**Comment by Terrance** [comment]
Follow-up on Live logins: turns out a fix has been found, which I've now applied to the `live-login-fix` branch.  This restores the use of `SkypeConnection.liveLogin(user, pwd)` to take the Live login flow, complete with the previous caveats (2FA remains not supported etc.).  If you were receiving the _Couldn't retrieve t field_ error before, please take a look and see if this solves it.

The `Skype` constructor uses `SkypeConnection.setUserPwd()` and `SkypeConnection.getSkypeToken()`, which have been updated to select SOAP via a new `SkypeConnection.soapLogin(user, pwd)` method.  This means the new (since 0.10) default of SOAP for login remains in place, though you can choose between them with an [Advanced connections](https://skpy.t.allofti.me/usage.html#advanced-connections)-like flow.  SkPy makes no attempt to detect which login flow would be preferred for a given account -- you should offer the choice to the user.

In summary:
* `Skype(user, pwd)` will continue to use SOAP
* `SkypeConnection.liveLogin(user, pwd)` will revert to a fixed Live login for compatibility with manual login flows
* Please let me know if the fixed Live flow still has t-field issues
